cluster/src/main/java/org/apache/iotdb/cluster/query/ClusterPlanExecutor.java

```diff
@@ -322,59 +324,82 @@ protected int getNodesNumInGivenLevel(PartialPath path, int level) throws Metada
       throw new MetadataException(e);
     }
 
-    // Here we append a ** to the path to query the storage groups which have the prefix as 'path',
-    // if path doesn't end with **.
+    // Here we append a ** to the path to query the storage groups which have the prefix as 'path'.
+    // If path doesn't have a **, we append one ** to the query path to avoid missing storage
+    // groups.
     // e.g. we have SG root.sg.a and root.sg.b, the query path is root.sg, we should return the map
     // with key root.sg.a and root.sg.b instead of an empty one.
-    PartialPath wildcardPath = path;
-    if (!wildcardPath.getMeasurement().equals(IoTDBConstant.MULTI_LEVEL_PATH_WILDCARD)) {
-      wildcardPath = wildcardPath.concatNode(IoTDBConstant.MULTI_LEVEL_PATH_WILDCARD);
+    boolean hasMultiLevelWildcard = false;
+    for (String node : path.getNodes()) {
+      if (node.equals(IoTDBConstant.MULTI_LEVEL_PATH_WILDCARD)) {
+        hasMultiLevelWildcard = true;
+        break;
+      }
+    }
+    Map<String, List<PartialPath>> sgPathMap;
+    if (!hasMultiLevelWildcard) {
+      Map<String, List<PartialPath>> extendedSgPathMap =
+          IoTDB.metaManager.groupPathByStorageGroup(
+              path.concatNode(IoTDBConstant.MULTI_LEVEL_PATH_WILDCARD));
+      sgPathMap = new HashMap<>(extendedSgPathMap.size());
+      // We should remove the last ** to restore the pattern to make it equivalent with the query.
+      for (Entry<String, List<PartialPath>> extendSgPaths : extendedSgPathMap.entrySet()) {
+        List<PartialPath> validPaths =
+            extendSgPaths.getValue().stream()
+                .map(PartialPath::getDevicePath)
+                .collect(Collectors.toList());
+        sgPathMap.put(extendSgPaths.getKey(), validPaths);
+      }
+
```

Comment:
Why is this necessary? If the query does not have a wildcard, we should match it strictly according to it like other queries.

Reply:
Thanks for your comments. As we discussed online, I will refine those codes in a reasonable way soon.